### PR TITLE
SM Cathedral final Boss fix

### PR DESCRIPTION
**Pull request by insunaa**
## 🍰 Pullrequest

This PR fixes the following issues with the final boss from SM Cathedral:
- Scarlet Commander Mograine does not engage in combat unless attacked
- Door to Whitemane's room doesn't close when the first encounter fails after killing only Whitemane
- Scarlet Commander Mograine does not use Divine Shield at all
- Whitemane doesn't run towards Mograine's corpse when casting resurrect
- Whitemane doesn't engage in combat after being summoned unless attacked

### Proof

- https://youtu.be/gCDMtHt-MsA?t=2083

### Issues

- None

### How2Test

- Create Level 40 Character
- Go into SM Cathedral
- Clear the room before the final boss
- engage the final boss

### Todo / Checklist

- [x] Make Whitemane not target a player after casting sleep
- [x] Make Whitemane face Mograine while casting resurrect

### CODE REVIEW

File: src/game/AI/ScriptDevAI/scripts/eastern_kingdoms/scarlet_monastery/boss_mograine_and_whitemane.cpp [hunk 1]
```diff
@@ -126,8 +128,11 @@ struct boss_scarlet_commander_mograineAI : public ScriptedAI
...
         Creature* pWhitemane = m_pInstance->GetSingleCreatureFromStorage(NPC_WHITEMANE);
-        if (pWhitemane && !pWhitemane->IsAlive())
+        if (pWhitemane && !pWhitemane->IsAlive()){
+            if (!(m_pInstance->GetData(TYPE_MOGRAINE_AND_WHITE_EVENT) == NOT_STARTED) || !(m_pInstance->GetData(TYPE_MOGRAINE_AND_WHITE_EVENT) == FAIL))
+                m_pInstance->SetData(TYPE_MOGRAINE_AND_WHITE_EVENT, FAIL);
             pWhitemane->Respawn();
+        }            
     }
 
     void DamageTaken(Unit* /*dealer*/, uint32& damage, DamageEffectType /*damagetype*/, SpellEntry const* /*spellInfo*/) override
```
Comment (killerwife) [inline src/game/AI/ScriptDevAI/scripts/eastern_kingdoms/scarlet_monastery/boss_mograine_and_whitemane.cpp:138]: I suggest removing this func per rework and use the new SetDeathPrevention paradigm and new hook JustPreventedDeath.

Comment (insunaa) [reply src/game/AI/ScriptDevAI/scripts/eastern_kingdoms/scarlet_monastery/boss_mograine_and_whitemane.cpp]: Done

**Comment by insunaa** [comment]
Also done :)

**Comment by insunaa** [comment]
Mograine now untargets the player and correctly faces mograine while resurrecting.

File: src/game/AI/ScriptDevAI/scripts/eastern_kingdoms/scarlet_monastery/boss_mograine_and_whitemane.cpp [hunk 2]
```diff
@@ -337,6 +379,15 @@ struct boss_high_inquisitor_whitemaneAI : public ScriptedAI
             DoCastSpellIfCan(m_creature, SPELL_DEEPSLEEP, CAST_INTERRUPT_PREVIOUS);
             m_bCanResurrectCheck = true;
             m_bCanResurrect = true;
+            if (Creature* pMograine = m_pInstance->GetSingleCreatureFromStorage(NPC_MOGRAINE))
+            {
+                SetCombatMovement(false);
+                m_creature->AttackStop(true);
+                float fX, fY, fZ;
+                pMograine->GetContactPoint(m_creature, fX, fY, fZ, INTERACTION_DISTANCE);
```
Comment (insunaa) [inline src/game/AI/ScriptDevAI/scripts/eastern_kingdoms/scarlet_monastery/boss_mograine_and_whitemane.cpp:387]: `INTERACTION_DISTANCE/2.f` might be more realistic, but can't test on reference

**Comment by insunaa** [comment]
Hmm...
https://classic.wowhead.com/npc=3977/high-inquisitor-whitemane#comments:id=3674:reply=294272
suggests that if mograine is dead and you wipe, whitemane resets to behind the door, mograine *doesn't* respawn and the door remains o pen.

Now I'm not sure anymore if either of them respawn on fail... Can anybody test this on Classic?

File: src/game/AI/ScriptDevAI/scripts/eastern_kingdoms/scarlet_monastery/boss_mograine_and_whitemane.cpp [hunk 3]
```diff
@@ -422,6 +427,12 @@ struct boss_high_inquisitor_whitemaneAI : public ScriptedAI
         else
             m_uiHolySmite_Timer -= uiDiff;
 
+        if(urand(1, 1000) == 1){
```
Comment (insunaa) [inline src/game/AI/ScriptDevAI/scripts/eastern_kingdoms/scarlet_monastery/boss_mograine_and_whitemane.cpp:430]: No idea what the actual values are, unfortunately, so I'm just giving it a 1:1000 chance of occurring. According to the comments it happens very rarely, and I think that should be given like this, but it's probably better to replace this with a known value, once it exists.

**Comment by killerwife** [comment]
LAst suggestion, I apologize for replying so late. Look up CombatAI and try to rewrite the script using it. Its our latest AI writing paradigm and it simplifies the core. However if you look at it and dont feel like it, thats fine, let me know, get me some debugging chat commands with which to test this, and what to look out for and I will try to merge it without it.

**Comment by insunaa** [comment]
Is there some simple example for CombatAI? I've seen a few examples, like the Shade of Akama script and there's quite a few things that I don't really understand (for example how AddCombatAction() works, etc.)

If there is some basic guide/tutorial/sample script that lays that out *reasonably* easily to understand, I'll be glad to rewrite this for CombatAI, but currently it looks very confusing to me

Also what do you mean debugging chat commands? As in what level the character needs to be, where the character needs to teleport to, etc.?

**Comment by insunaa** [comment]
@killerwife OK, I did an initial attempt at converting to CombatAI, but I'm running into a bit of trouble: The Divine shield timer of 40 seconds is supposed to pause wherever it's at when Mograine feigns death, and then resume once he wakes up.

Additionally all of the other CombatActions are supposed to stop while he's feigning death, but actually he's still casting them while lying on the ground. Any ideas?

https://github.com/insunaa/mangos-tbc/commit/223f7ecf549a386b596916f62a66aa5cefb07072

**Comment by killerwife** [comment]
There is a bool used for disabling any combat activity, however you must remember to set it and reset it yourself. SetCombatScriptStatus(true);

Also, which spell is being used for Feign Death? Its supposed to be a stun aura if I recall correctly, but maybe I am wrong.

**Comment by insunaa** [comment]
@killerwife 
It's not really a spell, it's just disabling combat movement, setting standstate to dead and making him unselectable etc.

I managed to make that part work now, but I still have trouble with the Divine Shield.

Also the timers seem wrong, because the spells sometimes take 3 times as long as was defined in AddCombatAction() (for example Hammer of Justice is supposed to come after 9.6s the first time, but it comes after ~22s)

**Comment by insunaa** [comment]
Making progress, but it's really janky so far...

**Comment by killerwife** [comment]
I checked his sniff, he indeed just uses UNIT_STAND_STATE_DEAD and when he is faking death he also has UNIT_FLAG_IMMUNE_TO_NPC and UNIT_FLAG_NOT_SELECTABLE.

**Comment by insunaa** [comment]
OK, this here should be a pretty complete CombatAI implementation. 1 bug that it has is that Whitemane turns around while resurrecting Mograine, I can probably fix this by adding another TimerlessCombatAction that ensures she faces Mograine at all times between casting `Deep Sleep` and re-entering combat, but I'm too tired to finish today.

https://github.com/insunaa/mangos-tbc/blob/sm_cath_fix_combatai/src/game/AI/ScriptDevAI/scripts/eastern_kingdoms/scarlet_monastery/boss_mograine_and_whitemane.cpp

**Pull request closed without merging**